### PR TITLE
types: Fix duration parser

### PR DESCRIPTION
**Pull request by iosmanthus**
Signed-off-by: Iosmanthus Teng <myosmanthustree@gmail.com>

### What problem does this PR solve? 
Fix `Duration` parser, make it compatible with `TiKV` as well as `MySQL`.

### Check List 

Tests 

 - [x] Unit test

Code changes

 - [x] Has exported function/method change

### CODE REVIEW

File: types/time.go [hunk 1]
```diff
@@ -30,6 +29,7 @@ import (
 	"github.com/pingcap/tidb/sessionctx/stmtctx"
 	"github.com/pingcap/tidb/util/logutil"
 	tidbMath "github.com/pingcap/tidb/util/math"
+	parser "github.com/pingcap/tidb/util/parser"
```
Comment (lonng) [inline types/time.go]: No need to use an alias for this package.

File: types/time_test.go [hunk 2]
```diff
@@ -251,7 +251,7 @@ func (s *testTimeSuite) TestTime(c *C) {
 		{"-838:59:59", "-838:59:59"},
 		{"838:59:59", "838:59:59"},
 		{"2011-11-11 00:00:01", "00:00:01"},
-		{"2011-11-11", "00:00:00"},
```
Comment (lonng) [inline types/time_test.go]: It's better to keep this case.

File: executor/write_test.go [hunk 3]
```diff
@@ -261,7 +261,7 @@ func (s *testSuite) TestInsert(c *C) {
 	// issue 6424
 	tk.MustExec("drop table if exists t")
 	tk.MustExec("create table t(a time(6))")
-	tk.MustExec("insert into t value('20070219173709.055870'), ('20070219173709.055'), ('-20070219173709.055870'), ('20070219173709.055870123')")
+	tk.MustExec("insert into t value('20070219173709.055870'), ('20070219173709.055'), ('20070219173709.055870'), ('20070219173709.055870123')")
```
Comment (lonng) [inline executor/write_test.go]: Insert some negative cases is proper (the first and the third case are same), it's better to change the L265.

**Comment by lonng** [comment]
Ref: https://github.com/tikv/tikv/pull/6492

File: executor/write_test.go (hunk 3 above)
Comment (iosmanthus) [reply executor/write_test.go]: If there is an invalid string for `Duration`, this `SQL` statement will fail so does the `tk.MustExec` function.

File: types/time_test.go (hunk 2 above)
Comment (iosmanthus) [reply types/time_test.go]: Have reverted this change.

File: util/parser/parser.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
+package parser
```
Comment (qw4990) [inline util/parser/parser.go]: The license should be added here.

File: util/parser/parser_test.go [hunk 5]
```diff
@@ -0,0 +1,156 @@
+package parser
```
Comment (qw4990) [inline util/parser/parser_test.go]: ditto

File: types/time.go [hunk 6]
```diff
@@ -1265,139 +1265,235 @@ func (d Duration) MicroSecond() int {
...
+	if negative {
 		d = -d
 	}
-
 	d, err = TruncateOverflowMySQLTime(d)
-	return Duration{Duration: d, Fsp: fsp}, errors.Trace(err)
+	return Duration{d, fsp}, errors.Trace(err)
+}
+
+func canFallbackToDateTime(str string) bool {
```
Comment (qw4990) [inline types/time.go]: Please add some comments for this function, for example in what cases we can fallback to `DateTime`.

File: util/parser/parser.go [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+		return unicode.IsSpace(rune(c))
+	}, times)
+	return rest, err
+}
+
+// Space0 matches at least 0 space.
+func Space0(buf string) string {
+	rest, err := Space(buf, 0)
+	if err != nil {
+		panic(err)
```
Comment (qw4990) [inline util/parser/parser.go]: It's better to return an error here instead of causing a panic?

File: types/time.go [hunk 8]
```diff
@@ -1265,139 +1265,235 @@ func (d Duration) MicroSecond() int {
...
+	rest = parser.Space0(rest)
+	overflow, frac, rest, err := matchFrac(rest, fsp)
+	if err != nil || len(rest) > 0 {
+		return ZeroDuration, ErrTruncatedWrongVal.GenWithStackByArgs("time", str)
+	}
+
 	if overflow {
-		second++
-		fracPart = 0
+		hhmmssAddOverflow(hhmmss[:], overflow)
```
Comment (qw4990) [inline types/time.go]: Why use `hhmmss[:]` here instead of `hhmmss` directly?

File: executor/write_test.go [hunk 9]
```diff
@@ -261,8 +261,8 @@ func (s *testSuite) TestInsert(c *C) {
 	// issue 6424
 	tk.MustExec("drop table if exists t")
 	tk.MustExec("create table t(a time(6))")
-	tk.MustExec("insert into t value('20070219173709.055870'), ('20070219173709.055'), ('-20070219173709.055870'), ('20070219173709.055870123')")
```
Comment (qw4990) [inline executor/write_test.go]: Why change this case?

File: util/parser/parser.go (hunk 7 above)
Comment (lonng) [reply util/parser/parser.go]: Seems it's a unreachable path.

File: util/parser/parser.go (hunk 4 above)
Comment (iosmanthus) [reply util/parser/parser.go]: Done.

File: util/parser/parser.go (hunk 7 above)
Comment (iosmanthus) [reply util/parser/parser.go]: `Space0` always successes.

File: executor/write_test.go (hunk 9 above)
Comment (iosmanthus) [reply executor/write_test.go]: If we try to insert a DateTime literal, the input should not carry a sign. This behaviour is `MySQL` compatible.

File: types/time.go (hunk 8 above)
Comment (iosmanthus) [reply types/time.go]: The type of `hhmmss` is `[3]int`, an array of `int` instead of a slice

File: types/time.go (hunk 6 above)
Comment (iosmanthus) [reply types/time.go]: Done.

**Pull request merged**